### ISSUE TITLE
asyncssh.connection.run() does not seem to respect timeout parameter

### ISSUE THREAD

**Issue by bkienker** [ronf/asyncssh#411]
I'm trying to allow the caller to specify a timeout value for SSH connections with asyncssh. Here's the function that creates the connection:

```
async def _run_client(host, command, timeout, **kwargs):
    async with asyncssh.connect(host, **kwargs) as conn:
        return await conn.run(command, timeout=timeout), host
```

The coroutine is gathered into a list of tasks in another function:

```
async def _run_multiple_clients(hosts, command, timeout, **kwargs):
    tasks = (_run_client(host, command, timeout, **kwargs) for host in hosts)
    results = await asyncio.gather(*tasks, return_exceptions=True)

    return results
```

And the tasks are run with a third function:

```
def run_ssh(hosts, command, timeout=60, **kwargs):
    return asyncio.get_event_loop().run_until_complete(
        _run_multiple_clients(hosts, command, timeout=timeout, **kwargs)
    )
```

In testing,  I've found that when I simulate a connection timeout by attempting to connect to a host that is behind a firewall, my function always takes a little over two minutes before raising the TimeoutError.

Should this timeout parameter affect the connection timeout wait time, or is it just for command execution time after the connection is established?

I'm sure I must be doing something wrong, thank you in advance. Please let me know if any additional information would be helpful.

**Comment by ronf** [comment]
The timeout in the run() method definitely only counts the time to run the command, as it doesn't even begin to execute until after the asyncssh.connect() call returns with a fully established (and authenticated) connection.

Even if you were to try and add a login_timeout to the connect() call, that will not begin to apply until after the SSHClientConnection object is created, which only happens after the TCP connection is established. So, it will allow you to time out if the SSH handshake takes too long but not if the TCP connect is taking a long time, like in your example here with a firewall.

If you want to enforce a timeout which includes the time to make the TCP connection, you'll need to use something like asyncio.wait_for() wrapped around whatever block of code you want the timeout to apply to. Take a look at #409 for more information on this.

**Comment by bkienker** [comment]
OK, that is extremely helpful! Thank you for your quick reply. I will take a look at asyncio.wait_for().

**Comment by ronf** [comment]
After multiple requests/questions about this, I've decided to add a new "connect_timeout" parameter that is basically shorthand for wrapping your connect call in asyncio.wait_for(). See #415 for more details about the new option, which is now available as commit b2cadab in the "develop" branch.

**Comment by bkienker** [comment]
That's excellent, thanks for your work!
